x/upgrade/module.go

```diff
@@ -0,0 +1,134 @@
+
+func init() {
+	RegisterCodec(moduleCdc)
+}
+
+var (
+	_ module.AppModule      = AppModule{}
+	_ module.AppModuleBasic = AppModuleBasic{}
+)
+
+// AppModuleBasic implements the sdk.AppModuleBasic interface
+type AppModuleBasic struct{}
+
+// Name returns the ModuleName
+func (AppModuleBasic) Name() string {
+	return ModuleName
+}
+
+// RegisterCodec registers the upgrade types on the amino codec
+func (AppModuleBasic) RegisterCodec(cdc *codec.Codec) {
+	RegisterCodec(cdc)
+}
+
+// RegisterRESTRoutes registers all REST query handlers
+func (AppModuleBasic) RegisterRESTRoutes(ctx context.CLIContext, r *mux.Router) {
+	rest.RegisterRoutes(ctx, r)
+}
+
+// GetQueryCmd returns the cli query commands for this module
+func (AppModuleBasic) GetQueryCmd(cdc *codec.Codec) *cobra.Command {
+	queryCmd := &cobra.Command{
+		Use:   "upgrade",
+		Short: "Querying commands for the upgrade module",
+	}
+	queryCmd.AddCommand(client.GetCommands(
+		cli.GetPlanCmd(StoreKey, cdc),
+		cli.GetAppliedHeightCmd(StoreKey, cdc),
+	)...)
+
+	return queryCmd
+}
+
+// GetTxCmd returns the transaction commands for this module
+func (AppModuleBasic) GetTxCmd(cdc *codec.Codec) *cobra.Command {
+	txCmd := &cobra.Command{
+		Use:   "upgrade",
+		Short: "Upgrade transaction subcommands",
+	}
+	txCmd.AddCommand(client.PostCommands()...)
+	return txCmd
+}
+
+// AppModule implements the sdk.AppModule interface
+type AppModule struct {
+	AppModuleBasic
+	keeper *Keeper
+}
+
+// NewAppModule creates a new AppModule object
+func NewAppModule(keeper *Keeper) AppModule {
+	return AppModule{
+		AppModuleBasic: AppModuleBasic{},
+		keeper:         keeper,
+	}
+}
+
+// RegisterInvariants does nothing, there are no invariants to enforce
+func (AppModule) RegisterInvariants(_ sdk.InvariantRegistry) {}
+
+// Route is empty, as we do not handle Messages (just proposals)
+func (AppModule) Route() string { return "" }
+
+// NewHandler is empty, as we do not handle Messages (just proposals)
+func (am AppModule) NewHandler() sdk.Handler { return nil }
+
+// QuerierRoute returns the route we respond to for abci queries
+func (AppModule) QuerierRoute() string { return QuerierKey }
+
+// NewQuerierHandler registers a query handler to respond to the module-specific queries
+func (am AppModule) NewQuerierHandler() sdk.Querier {
+	return NewQuerier(am.keeper)
+}
+
+// InitGenesis is ignored, no sense in serializing future upgrades
+func (am AppModule) InitGenesis(ctx sdk.Context, data json.RawMessage) []abci.ValidatorUpdate {
+	return []abci.ValidatorUpdate{}
+}
+
+// DefaultGenesis is an empty object
+func (AppModuleBasic) DefaultGenesis() json.RawMessage {
+	return []byte("{}")
+}
+
+// ValidateGenesis is always successful, as we ignore the value
+func (AppModuleBasic) ValidateGenesis(bz json.RawMessage) error {
+	return nil
+}
+
+// ExportGenesis is always empty, as InitGenesis does nothing either
+func (am AppModule) ExportGenesis(ctx sdk.Context) json.RawMessage {
+	return am.DefaultGenesis()
```

Comment:
Do we not need to export any genesis state as far as proposals and plans?

Reply:
If you do a hard-fork state dump, then you don't need/want these upgrade plans.

I can see them also being used with the current method as a forced halt time. Eg. you vote on upgrade plan with name "hub-4", when the time comes, it halts waiting for the binary with the handler (and inline migration script).

If you prefer, you can orchestrate a state export and fresh chain at that point. The fresh chain should not have the same upgrade plan, as the hard-fork did the change already. I see no point in them being in genesis, just used to change an existing, running change *without restart*.